File: docs/Sword2offer/CH2-编程题目-基础部分/Python/linkedListT2B.py
class Item:
  def __init__(self, value):
    self.value = value
    self.next = None

class LinkedList: 
  def __init__(self):
    self.head = None
  
  def append(self, value):
    if self.head == None:
      self.head = Item(value)
    else:
      item = Item(value)
      current = self.head
      while current.next:
        current = current.next
      current.next = item
  
  def __reverse_print(self, root):
    if root:
      if root.next:
        self.__reverse_print(root.next)
      print(root.value)

  def reverse_print(self):
    if not self.head:
      print('it\'s empty')
    else:
      self.__reverse_print(self.head)

if __name__ == '__main__':
  linkedList = LinkedList()
  linkedList.append(1)
  linkedList.append(2)
  linkedList.append(3)
  linkedList.append(4)
  linkedList.reverse_print()

File: docs/Sword2offer/CH2-编程题目-基础部分/ES6/linkedListT2B.js
let LinkedList = function () {
  let head = null
  let Item = function (value) {
    this.value = value
    this.next = null
  }
  // 递归方法变为私有方法
  function rPrint (root, callback) {
    if (root) {
      // 直到链表末尾才进行操作
      if (root.next) {
        rPrint(root.next, callback)
      }
      callback(root.value)
    }
  }
  this.push = (value) => {
    if (!head) {
      head = new Item(value)
    } else {
      let item = new Item(value),
        current = head
      // 找到末尾节点
      while (current.next) {
        current = current.next
      }
      current.next = item
    }
  }
  this.reversePrint = (callback) => {
    if (!head) {
      callback('it is empty')
    }
    rPrint(head, callback)
  }
}

let linkedList = new LinkedList()
linkedList.push(1)
linkedList.push(2)
linkedList.push(3)
linkedList.push(4)
linkedList.reversePrint(function (value) {
  console.log(value)
})
